### PR TITLE
Split strain into smaller frames

### PR DESCRIPTION
**Pull request by bhooshan-gadre**
If `frame-duration > 0`, it splits the conditioned strain into multiple smaller frame files of given `frame-duration`. If there is not enough data, then the last frame file will be of shorter duration than `frame-durarion`.

### CODE REVIEW

**Comment by titodalcanton** [comment]
@bhooshan-gadre it looks like you committed an unintended change to the README, can you revert it?

**Comment by bhooshan-gadre** [comment]
@titodalcanton I have removed README changes

**Comment by spxiwh** [comment]
So this might be a point to raise the distinction between *frames* and *frame files*. There might be case for writing the data as multiple *frames* in the same *frame file* (for example the O2 cleaned data would be much quicker to read if we supported that). This would require deeper changes, and this depends on what problem you're trying to solve here ...

**Comment by bhooshan-gadre** [comment]
@spxiwh I just wanted to add functionality to split a frame file with long data frame into multiple smaller frame files as it is taking much longer for `pycbc_live` to process and stream data. 
I didn't get your point how the adding functionality to write multiple frames would have help?

**Comment by spxiwh** [comment]
At the moment `pycbc_live` would have to read *all* the data from a frame file when it accesses it. If the data is stored as multiple frames in a single frame file, it only needs to read those frames overlapping the data. Or in short, you only read the data you need.

In HDF terms, rather than storing the data as *one* array in the file, and having to read all of it if you need to access it, store the data as *multiple* arrays and only read those ones that you need.

**Comment by bhooshan-gadre** [comment]
OK. I will try to put multiple frames in one frame file.

**Comment by bhooshan-gadre** [comment]
@spxiwh What structure should you propose? Does lalframe allow hierarchical structure like:
```file -- frame1 -- channels
     -- frame2 -- channels?```

**Comment by titodalcanton** [comment]
Hold on a second. Are we sure that PyCBC Live will be able to read a single frame file with multiple frames as fast as it reads many small *frame files*? I am not sure that is the case and we should know that before changing the patch that Bhooshan already proposed.

File: bin/pycbc_condition_strain [hunk 1]
```diff
@@ -56,6 +56,9 @@ parser.add_argument('--low-frequency-cutoff', type=float,
                     help='Provide a low-frequency-cutoff for fake strain. '
                          'This is only needed if fake-strain or '
                          'fake-strain-from-file is used')
+parser.add_argument('--frame-duration', type=int, default=0,
```
Comment (titodalcanton) [inline bin/pycbc_condition_strain]: Small pet peeve: remove `default=0` here, and replace the check on line 89 with `if args.frame_duration is None:`

**Comment by spxiwh** [comment]
I believe that multiple frames in one frame file is quicker than multiple smaller frames, but am not certain. This is maybe tied to a broader point that we should have functionality in our frame writing tools to more efficiently write long frame files by using multiple frames ... Although I'm not sure exactly how we'd do that.

**Comment by titodalcanton** [comment]
So it sounds like we should test the multiple-frames-per-file approach, and have another option in `pycbc_condition_strain` to output such a format. Something like `--multiple-files`: if given write many files with a single frame, if not given write a single file with many frames.

If that takes too long, though, I suggest we stick to the original approach, which I know works and is needed to do injections more efficiently.

**Comment by spxiwh** [comment]
Agreed, if speed is needed go with the simpler approach.

**Comment by bhooshan-gadre** [comment]
@spxiwh I do not know if it is possible. But this function, `FrameWrite` in `lalframe` which is being used doesn't allow writing multiple frames to one frame file:
```
int XLALFrameWrite(LALFrameH * frame, const char *fname)                        
{                                                                               
    // LALFrFile *frfile = NULL;                                                
    LALFrameUFrFile *frfile = NULL;                                             
    char tmpfname[FILENAME_MAX];                                                
                                                                                
    /* open temporary file */                                                   
    snprintf(tmpfname, sizeof(tmpfname), "%s.tmp", fname);                      
    frfile = XLALFrameUFrFileOpen(tmpfname, "w");                               
    if (!frfile)                                                                
        goto failure;                                                           
                                                                                
    /* write frame */                                                           
    XLALFrameUFrameHWrite(frfile, frame);                                       
                                                                                
    /* close temporary file */                                                  
    XLALFrameUFrFileClose(frfile);                                              
                                                                                
    /* rename file */                                                           
    rename(tmpfname, fname);                                                    
    return 0;                                                                   
                                                                                
  failure:     /* unsuccessful exit */                                          
    XLALFrameUFrFileClose(frfile);                                              
    /* TODO: remove tempfile */                                                 
    return -1;                                                                  
}
```
This writes a temperary file and renames it. Also, various stream routines also assume single frame per file.

**Comment by bhooshan-gadre** [comment]
@titodalcanton Fixed the pet peeve. But not with is not None. Let me know if it is not OK?

**Comment by bhooshan-gadre** [comment]
> Ok, in an ideal world if `--frame-duration` is zero (or less) it should throw an error, but this is good enough.

Added that.

**Comment by spxiwh** [comment]
I had a quick chat with Jolien about what the "many frames in one frame file" would require. Long story short: Stick with the original patch.

**Comment by bhooshan-gadre** [comment]
I had sent email to DASWG. But went to editors due to AEI email ID!

**Comment by spxiwh** [comment]
```
Yes, there are ways of doing this, but I think it requires a fairly low-level interface in the LAL implementation at the moment.

Here is an example program.  Is this what you’re looking for?

https://lscsoft.docs.ligo.org/lalsuite/lalframe/cat_8c_source.html

which makes use of some functions here:

https://lscsoft.docs.ligo.org/lalsuite/lalframe/utils_8c_source.html
```

**Comment by spxiwh** [comment]
(Also DASWG doesn't block AEI email addresses as such, it just blocks unrecognized email addresses similar to other ligo mailing lists)

File: bin/pycbc_condition_strain [hunk 2]
```diff
@@ -56,9 +56,15 @@ parser.add_argument('--low-frequency-cutoff', type=float,
                     help='Provide a low-frequency-cutoff for fake strain. '
                          'This is only needed if fake-strain or '
                          'fake-strain-from-file is used')
+parser.add_argument('--frame-duration', type=int,
+                    help='Split all data into smaller frame files of the given duration if specified.')
+
 pycbc.strain.insert_strain_option_group(parser)
 args = parser.parse_args()
 
+if args.frame_duration:
```
Comment (titodalcanton) [inline bin/pycbc_condition_strain]: Hmm, if it is zero, no error will actually be printed. I would use:
```python
if args.frame_duration is not None and args.frame_duration <= 0:
    parser.error('Frame duration should be positive integer, {} given'.format(args.frame_duration))
```
(there is an elegant way of doing this check "properly", but let's leave that for another time and close this issue).

**Comment by bhooshan-gadre** [comment]
@titodalcanton Sorry, I thought it is already merged. Found issue in `pycbc_make_psd_estimation_workflow`. Should I split this in 2 PRs?

**Comment by titodalcanton** [comment]
I am a bit lost about what is going on here. Why are changes to README and now `pycbc_make_psd_estimation_workflow` being made in this PR? Please just stick to the `pycbc_condition_strain` change and let's finalize it. Separate issues should go to different branches and different PRs.

**Comment by titodalcanton** [comment]
@bhooshan-gadre looking at the top of the page, it appears you are creating PRs by making the changes on the master branch of your git working copy. That is a risky workflow and it prevents you from working on different PRs at the same time. Usually you want to start a new PR like this:

* `git checkout master` ← make sure you base the change on the main branch
* `git pull` ← make sure you start from the current state of the main branch
* `git branch some_short_tag_explaining_the_change` ← branch off the current state
* `git checkout some_short_tag_explaining_the_change` ← switch to your new branch
* Make your change, `git add …`, `git commit`, repeat
* `git push your_remote_name some_short_tag_explaining_the_change` ← publish your branch on your github fork
* Create PR, get it approved and merged.
* `git branch -D some_short_tag_explaining_the_change` ← delete the branch once you are done

Hope this helps.

**Comment by titodalcanton** [comment]
Looks good.

**Pull request merged**